File: data structures and algorithms/java/06_038.java
import java.util.*;

/**
 * Graph implementation with Dijkstra's algorithm
 */
public class Graph {
   // Adjacency list representation
   private Map<String, List<Edge>> graph;
   
   /**
    * Edge representing connection between nodes
    */
   private static class Edge {
      String destination;
      int weight;
      
      public Edge(String destination, int weight) {
         this.destination = destination;
         this.weight = weight;
      }
   }
   
   /**
    * Initialize empty graph
    */
   public Graph() {
      this.graph = new HashMap<>();
   }
   
   /**
    * Add weighted edge to graph
    */
   public void addEdge(String fromNode, String toNode, 
         int weight) {
      // Check for negative weights
      if (weight < 0) {
         throw new IllegalArgumentException(
            "Negative weights not allowed in Dijkstra's");
      }
      
      // Ensure source node exists
      if (!graph.containsKey(fromNode)) {
         graph.put(fromNode, new ArrayList<>());
      }
      
      // Add the edge
      graph.get(fromNode).add(new Edge(toNode, weight));
      
      // Ensure destination node exists
      if (!graph.containsKey(toNode)) {
         graph.put(toNode, new ArrayList<>());
      }
   }
   
   /**
    * Find shortest paths using Dijkstra's algorithm
    */
   public Map.Entry<Map<String, Integer>, 
         Map<String, List<String>>> dijkstra(String start) {
      // Validate start node
      if (!graph.containsKey(start)) {
         throw new IllegalArgumentException(
            "Start node '" + start + "' not in graph");
      }
      
      // Initialize distances to infinity
      Map<String, Integer> distances = new HashMap<>();
      for (String node : graph.keySet()) {
         distances.put(node, Integer.MAX_VALUE);
      }
      distances.put(start, 0);
      
      // Priority queue for node processing
      PriorityQueue<Map.Entry<Integer, String>> pq = 
         new PriorityQueue<>(
            Comparator.comparing(Map.Entry::getKey)
         );
      pq.add(new AbstractMap.SimpleEntry<>(0, start));
      
      // Track paths to each node
      Map<String, List<String>> paths = new HashMap<>();
      List<String> startPath = new ArrayList<>();
      startPath.add(start);
      paths.put(start, startPath);
      
      // Process nodes by distance order
      while (!pq.isEmpty()) {
         // Get closest unprocessed node
         Map.Entry<Integer, String> current = pq.poll();
         int currentDistance = current.getKey();
         String currentNode = current.getValue();
         
         // Skip if we found better path already
         if (currentDistance > distances.get(currentNode)) {
            continue;
         }
         
         // Check all neighbors
         for (Edge edge : graph.get(currentNode)) {
            String neighbor = edge.destination;
            int weight = edge.weight;
            
            // Calculate total distance through current node
            int distance = currentDistance + weight;
            
            // Update if shorter path found
            if (distance < distances.get(neighbor)) {
               distances.put(neighbor, distance);
               
               // Create new path
               List<String> newPath = 
                  new ArrayList<>(paths.get(currentNode));
               newPath.add(neighbor);
               paths.put(neighbor, newPath);
               
               // Add to priority queue
               pq.add(new AbstractMap.SimpleEntry<>(
                  distance, neighbor));
            }
         }
      }
      
      return new AbstractMap.SimpleEntry<>(distances, paths);
   }
   
   /**
    * Example usage
    */
   public static void main(String[] args) {
      // Create a road network graph
      Graph g = new Graph();
      
      // Add edges (city connections with distances)
      g.addEdge("Dallas", "Chicago", 920);
      g.addEdge("Dallas", "Memphis", 410);
      g.addEdge("Chicago", "Boston", 850);
      g.addEdge("Memphis", "Atlanta", 335);
      g.addEdge("Memphis", "Chicago", 480);
      g.addEdge("Atlanta", "Miami", 610);
      g.addEdge("Atlanta", "Boston", 1070);
      g.addEdge("Boston", "Miami", 1450);
      
      // Find shortest paths from Dallas
      String startNode = "Dallas";
      Map.Entry<Map<String, Integer>, 
            Map<String, List<String>>> result = 
            g.dijkstra(startNode);
      
      Map<String, Integer> distances = result.getKey();
      Map<String, List<String>> paths = result.getValue();
      
      // Print distances
      System.out.println("Shortest distances from " + 
         startNode + ":");
      for (Map.Entry<String, Integer> entry : 
            distances.entrySet()) {
         System.out.println("  To " + entry.getKey() + 
            ": " + entry.getValue() + " miles");
      }
      
      // Print paths
      System.out.println("\nShortest paths:");
      for (Map.Entry<String, List<String>> entry : 
            paths.entrySet()) {
         String node = entry.getKey();
         List<String> path = entry.getValue();
         
         if (!node.equals(startNode)) {
            System.out.println("  To " + node + ": " + 
               String.join(" -> ", path) + 
               " (distance: " + distances.get(node) + 
               " miles)");
         }
      }
   }
}

File: data structures and algorithms/cs/06_038.cs
using System;
using System.Collections.Generic;

public class Graph
{
   private Dictionary<string, List<Edge>> graph;

   public Graph()
   {
      graph = new Dictionary<string, List<Edge>>();
   }

   public void AddEdge(
     string fromNode,
     string toNode,
     int weight
   )
   {
      if (weight < 0)
      {
         throw new ArgumentException(
           "Negative weights not allowed in Dijkstra's"
         );
      }

      if (!graph.ContainsKey(fromNode))
      {
         graph[fromNode] = new List<Edge>();
      }
      graph[fromNode].Add(new Edge(toNode, weight));

      if (!graph.ContainsKey(toNode))
      {
         graph[toNode] = new List<Edge>();
      }
   }

   public KeyValuePair<
     Dictionary<string, int>,
     Dictionary<string, List<string>>
   > Dijkstra(string start)
   {
      if (!graph.ContainsKey(start))
      {
         throw new ArgumentException(
           $"Start node '{start}' not in graph"
         );
      }

      var distances = new Dictionary<string, int>();
      foreach (var node in graph.Keys)
      {
         distances[node] = int.MaxValue;
      }
      distances[start] = 0;

      var pq = new PriorityQueue<
        KeyValuePair<int, string>,
        int
      >();
      pq.Enqueue(
        new KeyValuePair<int, string>(0, start),
        0
      );

      var paths = new Dictionary<string, List<string>>();
      var startPath = new List<string> { start };
      paths[start] = startPath;

      while (pq.Count > 0)
      {
         var current = pq.Dequeue();
         var currentNode = current.Value;
         var currentDist = current.Key;

         if (currentDist > distances[currentNode])
         {
            continue;
         }

         foreach (var edge in graph[currentNode])
         {
            var neighbor = edge.Destination;
            var weight = edge.Weight;
            var distance = currentDist + weight;

            if (distance < distances[neighbor])
            {
               distances[neighbor] = distance;
               var newPath = new List<string>(
                 paths[currentNode]
               ) { neighbor };
               paths[neighbor] = newPath;
               pq.Enqueue(
                 new KeyValuePair<int, string>(
                   distance,
                   neighbor
                 ),
                 distance
               );
            }
         }
      }

      return new KeyValuePair<
        Dictionary<string, int>,
        Dictionary<string, List<string>>
      >(distances, paths);
   }

   private class Edge
   {
      public string Destination { get; }
      public int Weight { get; }

      public Edge(string destination, int weight)
      {
         Destination = destination;
         Weight = weight;
      }
   }

   public static void Main(string[] args)
   {
      var g = new Graph();
      g.AddEdge("Dallas", "Chicago", 920);
      g.AddEdge("Dallas", "Memphis", 410);
      g.AddEdge("Chicago", "Boston", 850);
      g.AddEdge("Memphis", "Atlanta", 335);
      g.AddEdge("Memphis", "Chicago", 480);
      g.AddEdge("Atlanta", "Miami", 610);
      g.AddEdge("Atlanta", "Boston", 1070);
      g.AddEdge("Boston", "Miami", 1450);

      var startNode = "Dallas";
      var result = g.Dijkstra(startNode);
      var distances = result.Key;
      var paths = result.Value;

      Console.WriteLine(
        $"Shortest distances from {startNode}:"
      );
      foreach (var entry in distances)
      {
         Console.WriteLine(
           $"  To {entry.Key}: {entry.Value} miles"
         );
      }

      Console.WriteLine("\nShortest paths:");
      foreach (var entry in paths)
      {
         var node = entry.Key;
         var path = entry.Value;

         if (node != startNode)
         {
            var pathStr = string.Join(" -> ", path);
            Console.WriteLine(
              $"  To {node}: {pathStr} " +
              $"(distance: {distances[node]} miles)"
            );
         }
      }
   }
}
